docs/dev/testing.rst

```diff
@@ -0,0 +1,263 @@
+We use `Tox <https://tox.readthedocs.io>`__ to coordinate running different
+groups of tests together, and to make sure that the tests are run in a fresh
+Python environment without any other packages that might have been installed
+locally, but that won't necessarily be installed by PUDL users. Tox's behavior
+is configured with the ``tox.ini`` file in the main repository directory. There
+are several different "test environments" defined, to test different aspects
+of the software, or to perform other actions like building the documentation.
+To see a list of the available environments with short descriptions in the PUDL
+repository run:
+
+.. code-block:: console
+
+    $ tox -av
+
+    default environments:
+    ci               -> Run all continuous integration (CI) checks & generate test coverage.
+
+    additional environments:
+    flake8           -> Run the full suite of flake8 linters on the PUDL codebase.
+    pre_commit       -> Run git pre-commit hooks not covered by the other linters.
+    bandit           -> Check the PUDL codebase for common insecure code patterns.
+    linters          -> Run the pre-commit, flake8 and bandit linters.
+    doc8             -> Check the documentation input files for syntactical correctness.
+    docs             -> Remove old docs output and rebuild HTML from scratch with Sphinx
+    unit             -> Run all the software unit tests.
+    ferc1_solo       -> Test whether FERC 1 can be loaded into the PUDL database alone.
+    integration      -> Run all software integration tests and process a full year of data.
+    validate         -> Run all data validation tests. This requires a complete PUDL DB.
+    ferc1_schema     -> Verify FERC Form 1 DB schema are compatible for all years.
+    full_integration -> Run ETL and integration tests for all years and data sources.
+    full             -> Run all CI checks, but for all years of data.
+    build            -> Prepare Python source and binary packages for release.
+    testrelease      -> Do a dry run of Python package release using the PyPI test server.
+    release          -> Release the PUDL package to the production PyPI server.
+
+Continuous Integration Tests
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Our default tox test environment is ``ci`` -- which includes all of the tests
+that will be run in continuous integration using a
+`GitHub Action <https://github.com/features/actions>`__. You should run these
+tests before pushing code to the repository or making a pull request. Because it's
+the default test environment, it will be run if you call Tox without any arguments:
+
+.. code-block:: console
+
+    $ tox
+
+This is equivalent to:
+
+.. code-block:: console
+
+    $ tox -e ci
+
+If the dependencies have been changed, or you recently ran the tests while on
+another branch of the repository with other dependencies, you should tell Tox
+to rebuild the software environment it operates in with the ``-r`` (aka
+``--recreate``) flag:
```

Comment:
Will the tox tests fail if dependencies have been changed and you don't use the `-r` flag?

Reply:
@aesharpe If the dependencies have been changed, yes, it's likely something will fail if you don't recreate the environment (since it won't be aware of those new dependencies). However, we're much more likely to be editing our own code than changing our dependencies, and the environment setup takes a little while, which makes the process of testing more tedious -- especially if you're forcing it to recreate the environment every single time, and are just re-running a small part of the tests (like rebuilding the docs or something).

@cmgosnell When you say "the default" do you mean make that the first example of how to run a test with Tox? Or are you saying that `tox` by itself should automatically recreate the environment?  I don't know how to do the latter, or whether it's possible.

Reply:
I mean can we enable `tox -e ci` to run `-r`? don't you specify arguments for these tox tests in `tox.ini`? can `ci` just always run `-r` as well as whatever else it is running?

If not, than I think we should make the only example include `-r`.

Reply:
It looks like we can force any individual test environment to `recreate` if we want it to. This might make sense on the `ci` environment, since it already takes 20-30 minutes to run, so the extra time required (up to a couple of minutes) won't be a big deal.  Especially in the case where someone is trying to run the `unit` tests or build the docs or run the linters, it can end up being the majority of the time required to run the test. But probably users will still need/want to understand how this works.

Reply:
yeah I can imagine leaving it as an option for the unit tests but in the service of keeping it simple it makes sense to never require a users to think about this unless they want to

Reply:
I told it to recreate the environment every time it runs on `ci` and `full` and `validate` since those all take much longer to run, so the time required to set up the environment is small in comparison, and it would also be extra annoying if they failed because of environment out-of-sync stuff.

Reply:
great!